### PR TITLE
Extract more to ide module

### PR DESCRIPTION
**Pull request by imkiva**
@ice1000

### CODE REVIEW

**Comment by ice1000** [comment]
Can you also solve #795?

**Review by ice1000 — DISMISSED** [review]
Looks good

File: ide/src/main/java/org/aya/ide/action/Folding.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    var program = source.program().get();
+    if (program != null) program.forEach(folder);
+    return folder.foldingRanges.toImmutableSeq();
+  }
+
+  @Override public void accept(@NotNull Stmt stmt) {
+    switch (stmt) {
+      case Decl maybe -> Resolver.withChildren(maybe)
+        .filter(dv -> dv.concrete != null)
+        .map(dv -> dv.concrete.entireSourcePos())
```
Comment (ice1000) [inline ide/src/main/java/org/aya/ide/action/Folding.java:28]:
```suggestion
        .mapNotNull(dv -> dv.concrete)
        .map(dv -> dv.entireSourcePos())
```

And then use method ref

**Comment by imkiva** [comment]
> Can you also solve #795?

We will make `Reporter` and `Problem` generic types, is this intended?

**Comment by imkiva** [comment]
I would make `DistillerOptions` a generic type 😉

**Comment by ice1000** [comment]
> > Can you also solve #795?
> 
> We will make `Reporter` and `Problem` generic types, is this intended?

I think you can. Does that change a lot of existing code? 😟

**Comment by ice1000** [comment]
Well, I just do not think it's a good idea to have Aya-specific stuffs in the `tools` module. It should be just tools.

**Comment by imkiva** [comment]
bors d=ice1000

**Review by ice1000 — APPROVED** [review]
bors merge

**Pull request merged**